## Credit Preditor using Logistic Regression

### Setting up the environment

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn as sk
import numpy as np
import math

In [14]:
df = pd.read_csv("german_data.csv", sep=" ", header=None)
df.columns
df[20] = np.where(df[20] == 2, 0, 1)
df[20] = df[20].astype("category", copy=False)

In [15]:
df.head()
df.describe()


1             4            7            10           12  \
count  1000.000000   1000.000000  1000.000000  1000.000000  1000.000000   
mean     20.903000   3271.258000     2.973000     2.845000    35.546000   
std      12.058814   2822.736876     1.118715     1.103718    11.375469   
min       4.000000    250.000000     1.000000     1.000000    19.000000   
25%      12.000000   1365.500000     2.000000     2.000000    27.000000   
50%      18.000000   2319.500000     3.000000     3.000000    33.000000   
75%      24.000000   3972.250000     4.000000     4.000000    42.000000   
max      72.000000  18424.000000     4.000000     4.000000    75.000000   

                15           17  
count  1000.000000  1000.000000  
mean      1.407000     1.155000  
std       0.577654     0.362086  
min       1.000000     1.000000  
25%       1.000000     1.000000  
50%       1.000000     1.000000  
75%       2.000000     1.000000  
max       4.000000     2.000000

### Here it is important to NORMALIZE the values of the predictors

#### Normalizing (or scaling) the predictors helps when variables hugely vary in measurment (for example Age vs Salary),
#### allowing the model to give the same importance to every variable (independent of its units of measurment)
#### before the actual training process

In [26]:
X = df[[1,4,7,10,12,15,17]]##Only the numerical data, because transforming the categories to floating values would requiere pipelines
y = df[20]  ##Note that this column contains categories, not int64

In [34]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
## this is Z-score standarization, which basically substracts the mean to all the values to later devide each value by
## the standard deviation, usually making most values range from [-3;3] 
X_standarized = scaler.fit_transform(X)

### Preparing the data for training the model

In [28]:
from sklearn.model_selection import train_test_split

In [35]:
X_train, X_test, y_train, y_test = train_test_split(X_standarized, y, test_size=0.3, random_state=990)

### Introducing the model and training it

In [30]:
from sklearn.linear_model import LogisticRegression

In [31]:
md = LogisticRegression()
md.fit(X_train, y_train)
md.coef_

array([[-0.43688133, -0.07862818, -0.20658385, -0.03123329,  0.30904876,
         0.06984367, -0.06817791]])

### Trying for the predictions

In [32]:
predictions = md.predict(X_test)
predictions.view()

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1,
       1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

In [33]:
y_test.view()

C:\Users\botte\AppData\Local\Temp\ipykernel_9632\118822607.py:1: FutureWarning: Series.view is deprecated and will be removed in a future version. Use ``astype`` as an alternative to change the dtype.
  y_test.view()


707    0
725    1
91     1
19     1
31     1
      ..
698    1
307    0
776    1
257    0
297    1
Name: 20, Length: 300, dtype: category
Categories (2, int64): [0, 1]

### Evaluation of model

In [38]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, predictions)
print(f'Accuary: {accuracy: .2f}')

Accuary:  0.73


#### An accuary of 73% is actually incredible when we consider we didn't take into account most vaiables